cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -54,76 +169,274 @@ theorem not_contains_of_empty {α β} [BEq α] (k : α) :
 theorem make_wf [LT α] [StrictLT α] [DecidableLT α] (xs : List (α × β)) :
   WellFormed (Map.make xs)
 := by
-  simp only [WellFormed, make, toList, kvs, List.canonicalize_idempotent]
+  simp [WellFormed, make, toList, kvs, List.canonicalize_idempotent]
 
-theorem make_mem_list_mem [LT α] [StrictLT α] [DecidableLT α] {xs : List (α × β)} :
+/-
+  Note that the converse of this is not true:
+  counterexample `xs = [(1, false), (1, true)]`.
+  Then `Map.make xs = [(1, false)]`.
+-/
+theorem mem_kvs_make [LT α] [StrictLT α] [DecidableLT α] {xs : List (α × β)} :
   x ∈ (Map.make xs).kvs → x ∈ xs
 := by
   simp only [kvs, make]
   intro h₁
   have h₂ := List.canonicalize_subseteq Prod.fst xs
-  simp [List.subset_def] at h₂
+  simp only [List.subset_def] at h₂
   exact h₂ h₁
 
+/--
+  Very similar to `mem_kvs_make` above
+-/
+theorem mem_values_make [LT α] [StrictLT α] [DecidableLT α] {xs : List (α × β)} :
+  v ∈ (Map.make xs).values → v ∈ xs.map Prod.snd
+:= by
+  -- despite the similarity to `mem_kvs_make`, the proof does not currently
+  -- use `mem_kvs_make`
+  simp only [values, make]
+  simp only [List.mem_map, forall_exists_index, and_imp]
+  intro (k, v) h₁ h₂
+  exists (k, v)
+  subst h₂
+  simp only [and_true]
+  have h₂ := List.canonicalize_subseteq Prod.fst xs
+  simp only [List.subset_def] at h₂
+  exact @h₂ (k, v) h₁
+
 theorem make_nil_is_empty {α β} [LT α] [DecidableLT α] :
   (Map.make [] : Map α β) = Map.empty
 := by simp [make, empty, List.canonicalize_nil]
 
+/--
+  Note that the converse of this is not true:
+  counterexample `xs = [(1, false)]`, `ys = []`, `ab = (1, false)`.
+-/
+theorem make_cons [LT α] [DecidableLT α] {xs ys : List (α × β)} {ab : α × β} :
+  make xs = make ys → make (ab :: xs) = make (ab :: ys)
+:= by
+  simp only [make, mk.injEq]
+  apply List.canonicalize_cons
+
 /-! ### find? and mapOnValues -/
 
-theorem find?_mapOnValues {α β γ} [LT α] [DecidableLT α] [DecidableEq α] (f : β → γ) (m : Map α β) (k : α)  :
```

Comment:
Let's not change this signature.  Again, it will break internal proofs. 

Independently of that, if there are no preconditions to constrain the implicit variables (as in this case), they are usually made explicit.

Reply:
Can you explain more why we would make the implicit variables explicit here?  It seems to be more ergonomic to leave them implicit.  Lean is able to infer them at all of the current usage sites for this theorem.

Reply:
But I will put this change in a separate PR, reverting it here